experiments/composite/pkg/watchset/manager.go

```diff
@@ -0,0 +1,225 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+type Manager struct {
+	client     dynamic.Interface
+	restMapper meta.RESTMapper
+
+	mutex          sync.Mutex
+	interests      []*InterestSet
+	interestsByGVR map[schema.GroupVersionResource][]*InterestSet
+	watchesByGVR   map[schema.GroupVersionResource]*gvrWatcher
+}
+
+func NewManager(mgr manager.Manager) (*Manager, error) {
+	dynamicClient, err := dynamic.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, fmt.Errorf("error building dynamic client: %w", err)
+	}
+	m := &Manager{
+		client:         dynamicClient,
+		restMapper:     mgr.GetRESTMapper(),
+		interestsByGVR: make(map[schema.GroupVersionResource][]*InterestSet),
+		watchesByGVR:   make(map[schema.GroupVersionResource]*gvrWatcher),
+	}
+	return m, nil
+}
+
+func (w *Manager) newInterestSet(callback func()) *InterestSet {
+	interest := &InterestSet{
+		parent:   w,
+		byGVR:    make(map[schema.GroupVersionResource]*interestSetGVR),
+		callback: callback,
+	}
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	w.interests = append(w.interests, interest)
+
+	return interest
+}
+
+func (w *Manager) updateInterests() {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	watchsetsByGVR := make(map[schema.GroupVersionResource][]*InterestSet)
+
+	for _, interest := range w.interests {
+		if interest.closed {
+			// TODO: Clean up closed InterestSets from the slice
+			continue
+		}
+		for gvr := range interest.byGVR {
+			watchsetsByGVR[gvr] = append(watchsetsByGVR[gvr], interest)
+		}
+	}
+	w.interestsByGVR = watchsetsByGVR
+
+	// Make sure we are running watches for all interests
+	for gvr := range watchsetsByGVR {
+		watcher := w.watchesByGVR[gvr]
+		if watcher == nil {
+			klog.Infof("starting new watch for gvr %v", gvr)
+			watcher = &gvrWatcher{
+				gvr:        gvr,
+				parent:     w,
+				restMapper: w.restMapper,
+				client:     w.client,
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			watcher.cancel = cancel
+
+			go watcher.watchForever(ctx)
+
+			w.watchesByGVR[gvr] = watcher
+		}
+	}
+
+	// Close down any watches no longer of interest
+	for gvr, watcher := range w.watchesByGVR {
+		if _, found := watchsetsByGVR[gvr]; found {
+			continue
+		}
+
+		klog.Infof("stopping watch for gvr %v", gvr)
+		watcher.cancel()
+		delete(w.watchesByGVR, gvr)
+	}
+}
+
+func (w *Manager) onEvent(gvr schema.GroupVersionResource, ev *objectEvent) {
+	// TODO: rw-mutex
+	// TODO: register interests directly with watchers?
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	for _, interest := range w.interestsByGVR[gvr] {
+		interest.onEvent(gvr, ev)
+	}
+}
+
+type gvrWatcher struct {
+	client     dynamic.Interface
+	restMapper meta.RESTMapper
+
+	// TODO: Should we replace this with a direct pointer to the target(s)?
+	parent *Manager
+
+	// gvk schema.GroupVersionKind
+	gvr schema.GroupVersionResource
+
+	cancel func()
+}
+
+func (w *gvrWatcher) watchForever(ctx context.Context) error {
+	for {
+		if ctx.Err() != nil {
```

Comment:
Might be nice to also be able to break out on ctx.Done so we can actually implement a clean shut down. (Yeah I know we're currently using leader election but . . . )

Reply:
We do :-)  If the ctx is done, ctx.Err() is guaranteed to return non-nil.

I did replace time.Sleep(5 * time.Second) with a sleep-with-context function.